### ISSUE TITLE
Autosaves are created for unmodified files and not removed on close with split Editor panes

### ISSUE THREAD

**Issue by CAM-Gerlach** [spyder-ide/spyder#9265]
## Problem Description

I can confirm that with the latest ``master``, I'm still getting the ``Recover from autosave`` for files that were either saved or never unmodified when I closed Spyder (via the normal close button), with a clean ``autosave`` directory  and not other Spyders open before or currently. All the files I have open have autosaves created 1 minute after startup even if I've never touched Spyder, and they don't get removed when I close Spyder again. This happens consistently, every time with every file I either have open on startup or open subsequently, regardless of whether I view, modify, etc. it. 

I *finally* reproduced it starting from a clean ``--safe-mode`` Spyder instance by opening a two-pane layout ("horizontal split") and a few files and waiting 1 minute; it seems to be the two pane layout that triggers it, as it does not happen with only one pane open. I've tested enabling/disabling code style, docstring style analysis, projects,  and more, and none of them have an effect on whether it happens or not, just the two pane layout.

I also thought they should be removed by spyder-ide/spyder#8733 anyway, but apparently that handles it by merely removing the autosave when the "Don't save" button on the "Unsaved changes" dialog when closing the file is clicked, rather than when the file is actually closed (e.g. by the user, closing Spyder, switching projects, etc), and thus was not robust against any bugs (like that one) that might result in an autosave file being created for a reason other than an actual modification. 

### What steps reproduce the problem?

1. Open a clean copy of Spyder ``master``
2. Select "Split panes horizontally"
3. Open a file
4. Wait 60 seconds

## Versions

* Spyder version: ``master`` as of post time (``c90e2aea``)
* Python version: 3.7.2
* Qt version: 5.9.7
* PyQt version: 5.9.2
* Operating System name/version: Windows 8.1 x64

### Dependencies

```
pygments >=2.0    :  2.3.1 (OK)
qdarkstyle >=2.6.4:  2.6.5 (OK)
sphinx >=0.6.6    :  1.8.4 (OK)
pyls >=0.19.0     :  0.21.5 (OK)
nbconvert >=4.0   :  5.4.0 (OK)
pandas >=0.13.1   :  0.24.2 (OK)
numpy >=1.7       :  1.15.4 (OK)
sympy >=0.7.3     :  1.3 (OK)
cython >=0.21     :  0.29.4 (OK)
qtconsole >=4.2.0 :  4.4.3 (OK)
IPython >=4.0     :  7.2.0 (OK)
matplotlib >=2.0.0:  3.0.2 (OK)
pylint >=0.25     :  2.2.2 (OK)
```

**Comment by jitseniesen** [comment]
Great work, thanks. I don't think I ever tested two panes (because I never use it), so it makes sense that there are bugs lurking there.

**Comment by CAM-Gerlach** [comment]
> I don't think I ever tested two panes (because I never use it)

Really? I find Spyder basically unusable without two panes; so much wasted space and its much more efficient in my workflow to keep module code/functions in one, and my main script in another. But different strokes for different folks, as they say.

**Comment by goanpeca** [comment]
I barely use two panes as well :-p

**Comment by CAM-Gerlach** [comment]
Strange, I never use one pane :P It would be like going back to a non-multi-monitor setup once I've been used to having one for years and years, but I suppose we're all used to different things.

**Comment by CAM-Gerlach** [comment]
@jitseniesen @ccordoba12 Any chance this can get fixed for Beta 2? This leads to extremely annoying behavior making the otherwise excellent autosave and recovery features you spent so much time on nearly useless because it causes so much trouble (and is so hard to find any actually-desired autosaves in the morass of false ones) that as a user I'm strongly tempted to just turn it off.

**Comment by ccordoba12** [comment]
If you come up with a fix during the weekend, I'll merge it.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 @jitseniesen A workaround might be to add ``self.autosave.remove_autosave_file(finfo)`` [here](
![image](https://user-images.githubusercontent.com/17051931/57486247-b1bce000-727b-11e9-9be2-12a1e1826678.png)) 
to ``close_file()`` in ``editor.py``, which would remove any spuriously-created autosave files upon the file in question being successfully closed (and after handling the save/discard changes prompt). This wouldn't previous the files from being created, but it would clean them up if Spyder exits cleanly so they don't trigger a recovery prompt every time users launch Spyder. This isn't bad to do anyway, as it mostly mitigates any other bugs that might pop up with autosaves sticking around when they shouldn't, though it would be better to fix the initial problem.

 @jitseniesen , any thoughts on this?

**Comment by ccordoba12** [comment]
Sounds good. Please create a PR about it.

**Comment by goanpeca** [comment]
@CAM-Gerlach PR :-p ?

**Comment by CAM-Gerlach** [comment]
A few additional insights glaned from testing # :

**Comment by CAM-Gerlach** [comment]
*

**Comment by jitseniesen** [comment]
Sorry, I thoiught I had replied before. Go for it.

**Comment by CAM-Gerlach** [comment]
Sorry for the delay on #9327 @ccordoba12 @goanpeca and @jitseniesen ; it was much more complicated than anticipated and I had it finished except for the tests late last night, but I kept falling asleep working on them so I had to take a rest and finish in the morning.

Some insights gleaned from PR #9327 :

* Autosaves are created once following opening if more than one editor pane is open
* If the autosaves are removed after first being created, they won't be created again unless another pane is split
* If autosaves are deleted, or only one split pane is open, and then an additional pane is split, autosaves will be re-created (including if a panel is closed and then re-opened)
* Examining the ``name_mapping`` dict, items are present for all files, for all split panes except for the first one (index 0 in the Editorstack).
* Disabling all code, style, and docstring linting still produced the autosaves as before, even after the LSP server had stopped responding entirely and all LSP services were stopped.

**Comment by jitseniesen** [comment]
One bug is that when cloning an editor (e.g. because it appears in multiple panes), `.set_as_clone()` calls `.setDocument()` [here](https://github.com/spyder-ide/spyder/blob/540fa61f6125b35fd878ff6d6e86e3976ff52c55/spyder/plugins/editor/widgets/codeeditor.py#L623). This emits the `textChanged` signal which eventually leads to an autosave.

Another issue is that the autosave name mapping is stored on the `EditorStack` level and thus not synchronized between the different panes.

I have some ideas on how to fix this; I need to try out some things but I expect to have a PR ready this week.

**Comment by ccordoba12** [comment]
Thanks a lot @jitseniesen!

**Comment by CAM-Gerlach** [comment]
Fantastic @jitseniesen ! Looks like the former issue is the ultimate culprit, while the latter made the problem more difficult to detect and fix in PR #9327 (otherwise it would have been basically a one-line fix).
